python/pyarrow/_dataset.pyx

```diff
@@ -1421,34 +1447,93 @@ cdef class IsValidExpression(UnaryExpression):
         expr = make_shared[CIsValidExpression](operand.unwrap())
         self.init(<shared_ptr[CExpression]> expr)
 
+    def __reduce__(self):
+        return IsValidExpression, (self.operand,)
+
 
 cdef class CastExpression(UnaryExpression):
 
+    cdef CCastExpression *cast
+
     def __init__(self, Expression operand not None, DataType to not None,
                  bint safe=True):
-        # TODO(kszucs): safe is consistently used across pyarrow, but on long
-        #               term we should expose the CastOptions object
         cdef:
             CastOptions options
             shared_ptr[CExpression] expr
         options = CastOptions.safe() if safe else CastOptions.unsafe()
-        expr.reset(new CCastExpression(
-            operand.unwrap(),
-            pyarrow_unwrap_data_type(to),
-            options.unwrap()
-        ))
+        expr.reset(
+            new CCastExpression(
+                operand.unwrap(),
+                pyarrow_unwrap_data_type(to),
+                options.unwrap()
+            )
+        )
         self.init(expr)
 
+    cdef void init(self, const shared_ptr[CExpression]& sp):
+        UnaryExpression.init(self, sp)
+        self.cast = <CCastExpression*> sp.get()
+
+    @property
+    def to(self):
+        """
+        Target DataType or Expression of the cast operation.
+
+        Returns
+        -------
+        DataType or Expression
+        """
+        cdef:
+            shared_ptr[CDataType] typ = self.cast.to_type()
+            shared_ptr[CExpression] expr = self.cast.like_expr()
+        if typ.get() != nullptr:
+            return pyarrow_wrap_data_type(typ)
+        elif expr.get() != nullptr:
+            return Expression.wrap(expr)
+        else:
+            raise TypeError(
+                'Cannot determine the target type of the cast expression'
+            )
+
+    @property
+    def safe(self):
+        """
+        Whether to check for overflows or other unsafe conversions.
+
+        Returns
+        -------
+        bool
+        """
+        cdef CastOptions options = CastOptions.wrap(self.cast.options())
+        return options.is_safe()
+
+    def __reduce__(self):
+        return CastExpression, (self.operand, self.to, self.safe)
```

Comment:
IIUC CastExpression.__init__ must be modified as well (currently only takes a DataType `to`)

Reply:
Brbrbr, `InsertImplicitCasts` is only called from the Scanner's constructor on the Python side, so I think it is not possible to create a CastExpression with an Expression typed `to`.

@bkietz How about supporting only DataType argument on the python side?

Reply:
@kszucs Alright, leaving it for a follow up should be fine. Please create a jira and a matching TODO in the constructor (just in case someone needs to create a castlike and is confused why it doesn't work)

Reply:
Removed, so CastExpression only supports casting to datatypes (no cast like another expression functionality).

Created jira https://issues.apache.org/jira/browse/ARROW-8212